### PR TITLE
Add Value * == Value * peephole

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56966 [JIT] Disable Complete Shape Inlining For Testing Purposes
* #56828 Add unary/binary ops necessary for mobilenet
* #56642 Only optimize while change has been made
* #56438 [JIT] integer value refinement
* #56437 Factor out isDominatedBy
* **#55978 Add Value * == Value * peephole**
* #55977 [JIT] [Easy] Factor out non tensor peephole
* #55926 Add list len refinement
* #55925 Add handling of symbolic shapes
* #54809 Initial Symbolic Shape Analysis

This is needed for broadcasting two of the same symbolic shape

Differential Revision: [D27755328](https://our.internmc.facebook.com/intern/diff/D27755328)

### CODE REVIEW

File: torch/csrc/jit/passes/peephole_non_tensor.cpp [hunk 1]
```diff
@@ -114,6 +114,26 @@ struct PeepholeOptimizeNonTensorImpl {
...
+            node->inputs().at(0) != node->inputs().at(1)) {
+          continue;
+        }
+        auto inp_kind = node->inputs().at(0)->type()->kind();
+        // only handling common immutable types here because other types like
+        // Tensor or list of Tensor might throw on aten::eq
+        switch (inp_kind) {
+          case TypeKind::BoolType:
+          case TypeKind::IntType:
+          case TypeKind::FloatType: {
```
Comment (penguinwu) [inline torch/csrc/jit/passes/peephole_non_tensor.cpp:128]: What about `ComplexType` and `StringType`?

BTW, in which case would `x == x` throw exceptions on mutable types?

**Review by penguinwu — APPROVED** [review]
LGTM, just one minor question

File: torch/csrc/jit/passes/peephole_non_tensor.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/passes/peephole_non_tensor.cpp:128]: Tensor and anything containing Tensor

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27755328).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27755328).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27755328).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27755328).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27755328).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27755328).

**Pull request closed without merging**